crates/ink/codegen/src/generator/dispatch.rs

```diff
@@ -555,23 +561,52 @@ impl Dispatch<'_> {
                     }>>::IDS[#index]
                 }>>::CALLABLE
             );
+            let constructor_output = quote_spanned!(constructor_span=>
+                <#storage_ident as ::ink::reflect::DispatchableConstructorInfo<{
+                    <#storage_ident as ::ink::reflect::ContractDispatchableConstructors<{
+                        <#storage_ident as ::ink::reflect::ContractAmountDispatchables>::CONSTRUCTORS
+                    }>>::IDS[#index]
+                }>>::Output
+            );
             let deny_payment = quote_spanned!(constructor_span=>
                 !<#storage_ident as ::ink::reflect::DispatchableConstructorInfo<{
                     <#storage_ident as ::ink::reflect::ContractDispatchableConstructors<{
                         <#storage_ident as ::ink::reflect::ContractAmountDispatchables>::CONSTRUCTORS
                     }>>::IDS[#index]
                 }>>::PAYABLE
             );
+            let constructor_value = quote_spanned!(constructor_span=>
+                <::ink::reflect::ConstructorOutputValue<#constructor_output>
+                    as ::ink::reflect::ConstructorOutput::<#storage_ident>>
+            );
+
             quote_spanned!(constructor_span=>
                 Self::#constructor_ident(input) => {
                     if #any_constructor_accept_payment && #deny_payment {
                         ::ink::codegen::deny_payment::<
                             <#storage_ident as ::ink::reflect::ContractEnv>::Env>()?;
                     }
 
-                    ::ink::codegen::execute_constructor::<#storage_ident, _, _>(
-                        move || { #constructor_callable(input) }
-                    )
+                    let result: #constructor_output = #constructor_callable(input);
+                    let output_value = ::ink::reflect::ConstructorOutputValue::new(result);
+
+                    match #constructor_value :: as_result(&output_value) {
+                        ::core::result::Result::Ok(contract) => {
+                            ::ink::env::set_contract_storage::<::ink::primitives::Key, #storage_ident>(
+                                &<#storage_ident as ::ink::storage::traits::StorageKey>::KEY,
+                                &contract,
+                            );
+                            // On success we return the `Ok(())` value for callers.
+                            ::ink::env::return_value::<::core::result::Result<&(), ()>>(
+                                ::ink::env::ReturnFlags::default(), &::core::result::Result::Ok(&())
+                            )
+                        },
+                        ::core::result::Result::Err(err) => {
+                           ::ink::env::return_value::<::core::result::Result<(), & #constructor_value :: Error>>(
+                                ::ink::env::ReturnFlags::default(), &::core::result::Result::Err(err)
```

Comment:
Shouldn't we revert the transaction here **and** return `Error`?

Reply:
Indeed it absolutely should!